web/src/components/merchant/contexts/MerchantContext.tsx

```diff
@@ -0,0 +1,92 @@
+/**
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {NO_MERCHANT_WITH_FIREBASE_UID} from 'constants/errors/server-errors';
+import {USER_NOT_SIGNED_IN} from 'constants/errors/sign-in-errors';
+
+import React, {useContext, useState, Dispatch, SetStateAction} from 'react';
+
+import {getMerchantWithFirebaseUid} from 'api';
+import {getFirebaseUid} from 'firebase-auth';
+import {MerchantResponse} from 'interfaces';
+import {useHistory} from 'react-router-dom';
+
+type MerchantType = MerchantResponse | undefined;
+
+type ContextType =
+  | {
+      getMerchant: () => Promise<MerchantType>;
+      setMerchant: Dispatch<SetStateAction<MerchantType>>;
+    }
+  | undefined;
+
+const MerchantContext = React.createContext<ContextType>(undefined);
+
+/**
+ * useContext hook that ensures it is used within a MerchantContext.
+ */
+const useMerchantContext = () => {
+  const context = useContext(MerchantContext);
+  if (context === undefined) {
+    throw new Error(
+      'useMerchantContext must be used within a MerchantProvider'
+    );
+  }
+
+  return context;
+};
+
+/**
+ * MerchantContext provider with merchant getter and setter functions as its value.
+ */
+const MerchantProvider: React.FC = ({children}) => {
+  const [merchant, setMerchant] = useState<MerchantType>(undefined);
+  const history = useHistory();
+
+  // Returns merchant stored in the context state if not undefined, otherwise
+  // fetch merchant from server and update context state.
+  const getMerchant = async () => {
+    if (merchant === undefined) {
+      try {
+        const firebaseUid = await getFirebaseUid(); // Throws 'User not signed in' error.
+        const result = await getMerchantWithFirebaseUid(firebaseUid);
+        setMerchant(result);
```

Comment:
Your method title is called "getMerchant".
But in the code, your method does not seem to always about getting a merchant? (for example here you are setting merchant but not returning anything). The IF and ELSE flow are doing not semantically similar things. 

Consider either
1) Change the method name to reflect a universal meaning for all cases inside your method
or
2) Split them into two different methods, "getMerchant" and "UpdateMerchant".

Reply:
@xlai20 Actually, after we `setMerchant(result)` we will leave the `if` block and `return result`. That's why I didn't use early return.
So, the flow is something like, if merchant is not defined yet, then we fetch it first, updateMerchant and then return, otherwise, just return straight away. I think if this is not clear I can refactor the content of the if block into a separate function.